### DISCUSSION TITLE
How eagerload AbstractConcreteBase Polymorphic Model with Specific SQL Queries

### DISCUSSION THREAD

**Discussion by Maybe14** [sqlalchemy/sqlalchemy#11931] [Usage Questions]
I want to eagerly load a polymorphic model in a specific way. These are my current models:

```python
class Base(DeclarativeBase):
    id: Mapped[UUID] = mapped_column(postgresql.UUID(as_uuid=True), primary_key=True)
```

```python
class Left(Base):
    __tablename__ = "left"
    left_attribute1: Mapped[str]
    left_attribute2: Mapped[str]

    associations: Mapped[List["Association"]] = relationship(
        back_populates="left"
    )
    right_models: Mapped[List["RightBase"]] = relationship(
        secondary="association",
        primaryjoin="Left.id == Association.left_id",
        secondaryjoin="Association.right_id == RightBase.id",
        back_populates="left_models",
        viewonly=True
    )
```

```python
class Association(Base):
    __tablename__ = "association"
    other_attribute: Mapped[str]
   
    left_id: Mapped[UUID] = mapped_column(ForeignKey("left.id"))
    right_id: Mapped[UUID]

    left: Mapped["Left"] = relationship( 
        back_populates="associations"
    )
    right: Mapped["RightBase"] = relationship(
        back_populates="associations", 
        primaryjoin="and_(foreign(Association.right_id)==RightBase.id)" 
    )
```

```python
class RightBase(AbstractConcreteBase, Base):
    @declared_attr
    def associations(cls) -> Mapped[List["Association"]]:
        return relationship(
            "Association", 
            back_populates="right", 
            primaryjoin=lambda: and_(foreign(Association.right_id) == cls.id)
        )

    @declared_attr
    def left_models(cls) -> Mapped[List["Left"]]: 
        return relationship(
            "Left",
            secondary="association",
            primaryjoin=lambda: cls.id == Association.right_id,
            secondaryjoin=lambda: Association.left_id == Left.id,
            back_populates="right_models",
            viewonly=True
        )
        
class RightMixin:
    right_attribute1: Mapped[str]
    right_attribute2: Mapped[str]
    
class RightType1(RightMixin, RightBase):
    __tablename__ = "right_type_1"
    type = "type1"
   
    __mapper_args__ = {
         "polymorphic_identity": "type1",
         "concrete": True,
     }
    
class RightType2(RightMixin, RightBase):
    __tablename__ = "right_type_2"
    type = "type2"
   
    __mapper_args__ = {
         "polymorphic_identity": "type2",
         "concrete": True,
     }

Base.registry.configure()
```

In order to fetch the `right models` from the `left` eagerly, I do something like this:
```python
id_to_search_for = 12345
stmt = select(Left)
stmt = stmt.where(Left.id == id_to_search_for)
stmt = stmt.options(selectinload(Left.right_models))
result = await self.session.execute(stmt)
return result.scalars()
```

It "skips" the association table and returns the right models. This works fine and returns what I want, but upon looking at the SQL query log, it outputs something like this:
```sql
SELECT left_1.id AS left_1_id,
       pjoin.right_attribute1    AS pjoin_right_attribute1,
       pjoin.right_attribute2    AS pjoin_right_attribute2,
       pjoin.id                  AS pjoin_id,
       pjoin.TYPE                AS pjoin_type
FROM   left AS left_1
       join association AS association_1
         ON left_1.id = association_1.left_id
       join (SELECT right_type_1.right_attribute1  AS right_attribute1,
                    right_type_1.right_attribute2  AS right_attribute2,
                    right_type_1.id                AS id,
                    'type1'                        AS TYPE
             FROM   right_type_1
             UNION ALL
             SELECT right_type_2.right_attribute1  AS right_attribute1,
                    right_type_2.right_attribute2  AS right_attribute2,
                    right_type_2.id                AS id,
                    'type2'                        AS TYPE
             FROM   right_type_2
         ON association_1.right_id = pjoin.id
WHERE  left_1.id IN ( $1 :: uuid )
```

But I want it to look something like this:
```sql
SELECT left_1.id AS left_1_id,
       pjoin.right_attribute1    AS pjoin_right_attribute1,
       pjoin.right_attribute2    AS pjoin_right_attribute2,
       pjoin.id                  AS pjoin_id,
       pjoin.TYPE                AS pjoin_type
FROM   left AS left_1

JOIN   association AS association_1
ON     ON left_1.id = association_1.left_id

JOIN   right_type_1 AS pjoin_1
ON     association_1.right_id = pjoin_1.id

WHERE  left_1.id IN ( $1 :: uuid )

UNION ALL

SELECT left_1.id AS left_1_id,
       pjoin.right_attribute1    AS pjoin_right_attribute1,
       pjoin.right_attribute2    AS pjoin_right_attribute2,
       pjoin.id                  AS pjoin_id,
       pjoin.TYPE                AS pjoin_type
FROM   left AS left_1

JOIN   association AS association_1
ON     ON left_1.id = association_1.left_id

JOIN   right_type_2 AS pjoin_2
ON     association_1.right_id = pjoin_2.id

WHERE  left_1.id IN ( $1 :: uuid )
```

I've tried using the different relationship loaders, but I haven't found a combination that works. Is there a way to do this?

**Comment by zzzeek** [comment]
hi -

I've only skimmed briefly but it seems like your `select()` statement refers to a model called `EventMapping` that is not in your actual model.  This makes it *extremely difficult* for me to understand the geometry of your model and your query.   Concrete mappings are extremely difficult to work for relationships and this is why we try to discourage their use in the docs, so having to work around guessing what `EventMapping` represents in the model adds a lot of cognitive overhead for me.    Can you please provide a complete model **and** the query all at once to illustrate the full series of what you're trying to do ?   See the [mcve](https://github.com/sqlalchemy/sqlalchemy/issues?q=label%3A%22great+mcve%22+is%3Aclosed) tag for examples, thanks!

**Reply by Maybe14** [reply]
Hey, I just updated the post. It was supposed to be `left.right_models`. Thank you!

**Reply by zzzeek** [reply]
OK.  again just briefly, your base class for concrete loading is defined in terms of a union, "SELECT... FROM a UNION ALL SELECT.. FROM b".   the ORM sees the class as mapped to that construct, and it's "sealed" - the ORM can use that set of rows as the core of new statements, but it *cannot break into it and rearrange it*, since the ORM assumes nothing about how the mapped selectable that represents the rows is structured.

the two queries you show are potentially equivalent and there is the theoretical optimization of the query being rearranged either before being emitted to the database, or the database planner itself reducing both queries to the same query plan.    Overall I'd start with the latter and take a look at what your query planner is coming up with, because it might not make a difference.

**Comment by CaselIT** [comment]
Hi,

The example uses a selectinload while the sql seems to be of a joinedload.
-- edit: no that's selectin on a many to many. There is an issue for an optimization here https://github.com/sqlalchemy/sqlalchemy/issues/5987 but it will not create the query you are looking for 

In any case that's how the eager loaders are implemented. If you want to control the emotted query you have to manually construct it snd then use `contais_eager` to tell sqlalchemy to load the relationship from the query

**Reply by zzzeek** [reply]
it's not a joinedload, that's how the ORM SELECTs rows from a mapping against a union, it SELECTs from the query as a subquery.
